File: csharp/structural/facade.cs
namespace Gof.ModernCSharp.Structural;

// Facade hides subsystem details behind a smaller, easier entry point.
// Modern C# note: often just a clean service, client, or application-layer API.
public static class Facade
{
    sealed class Amplifier { public string On() => "amp on"; }
    sealed class Projector { public string On() => "projector on"; }
    sealed class Player { public string Play(string movie) => $"playing {movie}"; }

    sealed class HomeTheater
    {
        readonly Amplifier _amp = new();
        readonly Projector _projector = new();
        readonly Player _player = new();

        public string WatchMovie(string movie) => $"{_amp.On()}, {_projector.On()}, {_player.Play(movie)}";
    }

    public static string Run() => new HomeTheater().WatchMovie("Casablanca");
}

File: py/structural/facade.py
# Facade hides subsystem details behind a smaller, easier entry point.
# Modern Python note: this is often just a well-designed module, service, or client
# API that keeps messy subsystem coordination out of callers.
class Amplifier:
    def on(self):
        return "amp on"


class Projector:
    def on(self):
        return "projector on"


class Player:
    def play(self, movie):
        return f"playing {movie}"


class HomeTheaterFacade:
    def __init__(self):
        self.amp = Amplifier()
        self.projector = Projector()
        self.player = Player()

    def watch_movie(self, movie):
        return [self.amp.on(), self.projector.on(), self.player.play(movie)]


if __name__ == "__main__":
    print(HomeTheaterFacade().watch_movie("Casablanca"))
